### PR TITLE
Add support for colon separated sub-parameters

### LINKED ISSUE #4321: Support colons as a VT parameter separator
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.19041.21]
Windows Terminal version (if applicable): 0.8.10091.0

Any other software?  WSL, Ubuntu
```

# Steps to reproduce

I have TERM set to xterm-24bit using the standard termconfig instructions that are online.  Running emacs in terminal mode (emacs -nw) and then using M-x list-display-colors

# Behavior (windows terminal on the left, wsltty on the right)

image

In both cases, emacs knows about all 500+ colors available, but wsltty shows them and windows terminal doesn't.  Also this has been like this for a couple months at least, but now it's bugging me enough to ask for a fix. :)

### PR DESCRIPTION
**Pull request by tusharsnn**
Adds support for colon `:` separated sub parameters in parser. Technically, after this PR, nothing should change except, now sub parameters are parsed, stored safely and we don't invalidate the whole sequence when a `:` is received within a parameter substring.

In this PR:
- If sub parameters are detected with a parameter, but the usage is unrecognised, we simply *skip* the parameter in `adaptDispatch`. 
- A separate store for sub parameters is used to avoid too many changes to the codebase.
- We currently allow up to `6` sub parameters for each parameter, extra sub parameters are *ignored*.
- Introduced `VTSubParameters` for easy access to underlying sub parameters.

> **Info**: We don't use sub parameters for any feature yet, this is just the core implementation to support newer usecases.

## Validation Steps Performed
- [x] Use of sub parameters must not have any effect on the output.
- [x] Skip parameters with unexpected set of sub parameters.
- [x] Skip sequences with unexpected set of sub parameters.

References #4321
References #7228
References #15599
References https://github.com/xtermjs/xterm.js/pull/2751
Closes #4321

### CODE REVIEW

**Comment by tusharsnn** [comment]
It's still in draft, but I'll like someone to review and suggest changes if there needs to be any.

File: src/terminal/parser/stateMachine.cpp
Comment (zadjii-msft) [inline src/terminal/parser/stateMachine.cpp:1191]: this comment seems wrong? Er, I guess `\x1b[:m` is invalid - is it?

Comment (zadjii-msft) [reply src/terminal/parser/stateMachine.cpp:1191]: (might just need clarification)

File: src/terminal/adapter/DispatchTypes.hpp
Comment (zadjii-msft) [inline src/terminal/adapter/DispatchTypes.hpp:244]: the like, smallest nit ever: I might rename this to `_subParamRanges`. For some reason, it makes more sense in my head if you flip which nolun is plural. Somhow that makes it more clear that it's a set of ranges that contains the subparams for each param. 

Feel free to ignore me though :relaxed:

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp:1191]: I think, I tried to keep the existing behavior here. Should we allow such sequences? I'd be happy make the changes.

Sub parameters to empty parameter shouldn't have any effect, AFAIK.

File: src/terminal/adapter/DispatchTypes.hpp
Comment (tusharsnn) [reply src/terminal/adapter/DispatchTypes.hpp:244]: Yeah, it makes sense.. I'll update the PR. Thanks.

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: done. 👍

File: src/terminal/adapter/DispatchTypes.hpp
Comment (tusharsnn) [reply src/terminal/adapter/DispatchTypes.hpp]: made the requested changes👍

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp:1191]: > I think, I tried to keep the existing behavior here. Should we allow such sequences?

@zadjii-msft Should I make any changes here?

**Comment by j4james** [comment]
I have some issues with the standards compliance of this implementation.

Per STD 070, section 3.5.3.1, "The bit combination 3/10 is reserved for future standardization. If 3/10 is received within a parameter string, the entire sequence up to and including the final character shall be ignored." If we accept the ODA documentation as the "future standardization", that only defines 3/10 usage in the `SGR` sequence.[^footnote] Outside of that the original rules still apply.

So something like `\e[10:20H` should still be a no-op. It should _not_ be interpreted as `\e[10H`.

It's arguable that colons _could_ now be accepted in controls with selective parameters, but even then, any parameter substring that we don't yet support is the equivalent of an unimplemented parameter. And section 3.5.1.3 says "unimplemented selective parameter values of control sequences shall be ignored as if that parameter were not received."

So something like `\e[?5:10h` should also be no-op. It should _not_ be interpreted as `\e[?5h`.

The only thing up for debate is whether something like `\e[?10:10;5h` should be entirely ignored, or we could just ignore the initial parameter substring. But I'm leaning towards ignoring the entire thing until parameter substrings have defined usage in that particular control sequence.

Then within `SGR` itself, any parameter substrings that we don't yet recognise are still the equivalent of unimplemented parameters, so they should still be ignored as mentioned above.

So something like `\e[31:42m` should be a no-op. It should _not_ be interpreted as `\e[31m`. Even something like `\e[4:3m` should be a no-op, until we actually support extended underline sequences.

[^footnote]: Although if we're supporting parameter substrings in `SGR`, we probably also need to support them in the equivalent attribute parameters of `DECCARA` and `DECRARA`. But that may introduce another level of complexity, because a colon in any of the numeric parameters should still cause the sequence to be ignored - it should only be permitted within the selective attribute parameters.

**Comment by j4james** [comment]
> @j4james, what is ODA?

Sorry, I should have been clearer about that. By _ODA_, I meant the _Open Document Architecture_ format, ITU-T recommendation T.416 or ISO/IEC 8613-6. That's where modern terminals got the idea for the colon-based 256-color and RGB extensions, although it's not actually a terminal standard as such.

**Comment by tusharsnn** [comment]
@j4james I don't want to pretend that I understood everything what you said there, because I'm too new to all the inner workings of terminal. But I guess what you mean is, *incase we don't know a parameter accepts sub parameters or not, then we should better enter into `CsiIgnore` state, instead of simply parsing it further*. So it's more like a **opt-in to accepts sub parameter**, isn't it? 
I like this idea, this ensures future proofing, since if suddenly there's a new sequence(that uses sub parameter) and we don't how it works, it would safely be discarded on PTY layer.

Let me see how we can achieve it on this PR. Should that address all your concerns?

**Comment by j4james** [comment]
> So it's more like a **opt-in to accepts sub parameter**, isn't it?

@tusharsnn That's exactly right, yes. But note that the `CsiIgnore` state is currently handled at the `StateMachine` level, but you don't know what the actual control sequence is going to be used for until you've parsed the _final_ character, and then moved on to the `OutputStateMachineEngine`.

My initial idea was you could have a check at the start of the `ActionCsiDispatch` method in `OutputStateMachineEngine`, and that could immediately abort if the `parameters` contained any sub parameters, while the `id` was anything other than `SGR`. You'd also need a similar check in `ActionDcsDispatch`, except that should always abort if there are any sub parameters.

What I'm not sure about is how we'd handle `DECCARA` and `DECRARA`, because only some of their parameters are allowed to have sub parameters. I think that's going to end up being a bit messy.

**Comment by tusharsnn** [comment]
> But note that the CsiIgnore state is currently handled at the StateMachine level, but you don't know what the actual control sequence is going to be used for until you've parsed the final character, and then moved on to the OutputStateMachineEngine.

I'm toying with this idea:
```c++
// DispatchTypes.hpp

// we use this in VT parser to indicate that the parameter (GraphicsOption)
// accepts sub-parameters.
enum class GraphicsOptionsWithSubParams: VTInt
{
    Underline = 4,
};

// in stateMachine.cpp

void StateMachine::_EventCsiParam(const wchar_t wch)
{
...
else if (_isSubParameterDelimiter(wch))
    {
        if(_acceptsSubParam(_parameters.back()))
        {
            _ActionSubParam(wch);
            _EnterCsiSubParam();
        }
        else
        {
            _EnterCsiIgnore();
        }
    }
...
}
```

**Comment by tusharsnn** [comment]
> What I'm not sure about is how we'd handle DECCARA and DECRARA, because only some of their parameters are allowed to have sub parameters. I think that's going to end up being a bit messy.

@j4james Can you share a reference, so that I can have a look which parameters support sub parameters.

**Comment by j4james** [comment]
> ```c++
> if(_acceptsSubParam(_parameters.back()))
> ```

But you can't tell whether a given parameter accepts sub parameters at this point. For example, if the parameter is a `4`, it could indicate the underline attribute of an `SGR` sequence, but it could also be the row number 4 from a `CUP` sequence, or the insert-replace mode number in an `SM` or `RM` sequence, or a thousand other things.

> Can you share a reference, so that I can have a look on which parameters support sub parameters.

See https://vt100.net/docs/vt510-rm/DECCARA.html and https://vt100.net/docs/vt510-rm/DECRARA.html.

The first four parameters are numeric parameters, which can't have sub parameters. The remaining parameters are rendition attributes, which are selective parameters, and should theoretically support the same values that we allow in an `SGR` sequence.

**Comment by tusharsnn** [comment]
@j4james 
> But you can't tell whether a given parameter accepts sub parameters at this point. 

yeah, you're right. I realised that after I made the reply. right now, I think we should guard against both parameter and sequence to never allow parameter when we don't accept them. 

So, something like `\x1b[38:2::3m` should rightfully enter *CsiIgnore* right after we parse a `:` after 38. This should avoid the need to check *if a parameter has a sub parameter(s) when it isn't suppose to have one* for every parameter in `SGR` .

**Comment by tusharsnn** [comment]
Made few changes:
- `\e[4:3m` should no longer be interpreted as `\e[4m`
- Sequences like `\e[4;38:2m` will be discarded, underline is *not* applied. This aligns with the current state of the parser.
- `\e[10:20H` and `\e[?5:10h` is a no-op now.

@j4james 
> What I'm not sure about is how we'd handle DECCARA and DECRARA

Does the changes address this issue?

**Comment by tusharsnn** [comment]
I don't understand why test is failing? Is it again #8495 plaguing the CI? @zadjii-msft

File: src/terminal/parser/OutputStateMachineEngine.cpp
Comment (j4james) [inline src/terminal/parser/OutputStateMachineEngine.cpp:1116]: If we want to support sub params in the `DECCARA` and `DECRARA` sequences, this will need to be a little more complicated. In pseudocode, something like this:

```
return id == SGR || (firstSubParam >= 4 && (id == DECCARA || id == DECRARA));
```

Because we want to allow sub params in `DECCARA` and `DECRARA`, but only from parameter 4 onwards, since the first 4 parameters are numeric. I'm hoping it shouldn't be too complicated to track something like `firstSubParam` or an equivalent test.

File: src/terminal/parser/stateMachine.cpp
Comment (j4james) [reply src/terminal/parser/stateMachine.cpp:565]: It just occurred to me that it might be easier if you just set the `_parameterLimitReached` flag here as well. That should solve the problem without having to make major changes to your implementation.

So if someone happens to use 32 sub parameters as part of the first parameter, they won't get any additional parameters accepted at all. Dropping excessive parameters is fine - I just don't want them to be misinterpreted.

File: src/terminal/adapter/adaptDispatchGraphics.cpp
Comment (tusharsnn) [reply src/terminal/adapter/adaptDispatchGraphics.cpp]: added the requested changes.

File: src/terminal/parser/stateMachine.cpp
Comment (j4james) [reply src/terminal/parser/stateMachine.cpp:1205]: In the specific context of `SGR`, the default parameter value is 0, so I'd consider `:` to be a sub parameter of 0 (essentially the equivalent of `0:0`). Since that doesn't have a defined meaning that we're aware of, it classifies as an unimplemented selective parameter, which should be ignored (per STD 070 §3.5.1.3).

So I'd say `printf "\e[:;4mhello\n"` _should_ render as underlined. The first parameter is ignored, but the second is still executed.

I wouldn't take much notice of what other terminals are doing unless we can find some spec text that backs up their interpretation, because I don't think they've put much thought into this other than doing whatever was easiest for them. I've tested about a dozen of them, and _none_ of them interpret sub parameters in exactly the same way.

File: src/terminal/parser/stateMachine.cpp
Comment (j4james) [reply src/terminal/parser/stateMachine.cpp]: Let says you have an `SGR` sequence with these three parameters:

* `38:2::128:0:128::::::::::::::::::::::::::::`
* `4:0`
* `5`

Ideally all three would be executed, so you get magenta text, underline off, and blinking.

However, I wouldn't mind if we chose to ignore the second two, because the first one exceeded our global parameter limit. It's expected that a terminal might drop excessive parameters, so behavior like this would not be particularly surprising.

Dropping the second one and still executing the third seems a bit weird though. As a user that would look like a bug to me, because it's not at all obvious why that is happening. I don't feel that strongly about it, though, especially since this is never likely to happen in practice.

The only thing I _am_ strongly opposed to is having the second parameter interpreted as `4` instead of `4:0`. So as long as we don't do that, I'm happy.

Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: > Ideally all three would be executed, so you get magenta text, underline off, and blinking.

Since these properties are isolated from each other, I'd expect them to work when one of them is not in its standard format (this is already something which application should be blamed for).

I guess, I'd pick a terminal that can still parse `5` (and anything it's capable of, safely) when an application is throwing a bunch of `38:2::128:0:128::::::::::::::::::::::::::::` at it 😄 . It might make our users think, WT is able to parse at least something from a broken `SGR`/`DECCARA`/`DECRARA` sequence its receiving. But again, this is just my opinion 😅 .

It makes me wonder, should we block further parameter only because we couldn't handle sub parameters of a different parameter than we're currently on? We allow up to 32 parameters in a sequence, however, if we starts blocking them, it can come down to 1, if a malformed sub parameters list is received.

> I don't feel that strongly about it, though, especially since this is never likely to happen in practice.

Same. 👍

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp:1205]: > So I'd say printf "\e[:;4mhello\n" should render as underlined. The first parameter is ignored, but the second is still executed.

I think this can easily be done now, since we have `CsiParamIgnore` to ignore one parameter at a time, or ignore everything up to ";" and enter `CsiParam` afterwards.

File: src/terminal/parser/stateMachine.cpp
Comment (j4james) [reply src/terminal/parser/stateMachine.cpp]: > Since these properties are isolated from each other, I'd expect them to work when one of them is not in its standard format

But in this case _all_ of them are technically valid. It's only our internal implementation that is restricting how many of them we can parse. And that's fine, because the standard allows for terminals to drop parameters past a certain limit, but the expectation is that you're going to keep the first N parameters - we can't just decide we're going to keep the first and the third, but drop the second one.

Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: > but the expectation is that you're going to keep the first N parameters - we can't just decide we're going to keep the first and the third, but drop the second one.

That seems reasonable to me, let's refactor the code to behave this way. 👍

File: src/terminal/parser/OutputStateMachineEngine.cpp
Comment (tusharsnn) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: I think, `DECCARA` `DECRARA` is now handled properly. I'll wait for your confirmation.

File: src/terminal/adapter/adaptDispatchGraphics.cpp
Comment (j4james) [inline src/terminal/adapter/adaptDispatchGraphics.cpp:290]: I just realised that this sub parameter test is in the wrong place. It needs to be a level down inside the `_ApplyGraphicsOption` method, because that gets called directly by the `DECRARA` implementation. As it is now, `DECRARA` won't get the sub parameter support.

File: src/terminal/parser/OutputStateMachineEngine.cpp
Comment (j4james) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: Yeah. This part looks good to me. But see my comments below regarding the location of the `hasSubParamsFor` test.

File: src/terminal/parser/stateMachine.cpp
Comment (j4james) [inline src/terminal/parser/stateMachine.cpp:579]: I'm not sure that removing the last parameter here is the right approach. Because that means a sequence like this wouldn't work:

```
printf "\e[38:2::128:0:128:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1mMagenta?\e[m\n"
```

But this semicolon equivalent _does_ work, despite having the same limit:

```
printf "\e[38;2;128;0;128;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1mMagenta?\e[m\n"
```

Maybe it should only remove the last parameter if we haven't been able to add _any_ sub parameters yet? I guess that would be something like `_subParameterRanges.back().first >= MAX_SUBPARAMETER_COUNT`.

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: I guess this is addressed now. Lmk if it needs changes.

**Comment by tusharsnn** [comment]
> Maybe it should only remove the last parameter if we haven't been able to add any sub parameters yet?

Why should we make this is an exception to our previous rule which says `\e[4:0m` shouldn't be interpreted as `\e[4m`.

Let's say we have `38:2::128:0:128` to parse. We might interpret this as `38:2::128`, `38:2::128:0`, or `38:2::128:0:128` depending on how many spaces are left inside sub parameter list. This could lead to different colours all the time.

I kinda dislike the idea to not acknowledging the sub parameters send by application. The safest option is if we can't handle all them, we should just drop the parameter.

Assuming our own in place of discarded sub parameters is **very** unsafe in my opinion, because **it can render differently every time**, which might seem like a bug.

File: src/terminal/adapter/adaptDispatchGraphics.cpp
Comment (tusharsnn) [reply src/terminal/adapter/adaptDispatchGraphics.cpp]: done.

**Comment by tusharsnn** [comment]
There's still an edge case I need to figure out, `\e[1:<S1>:<S2>...:<S32>m` (first parameter overflowed the limit) shouldn't be interpreted as `\e[m`. 

Corrective action could be to add back `defaultParameter` in case we are removing first parameter. So `\e[1:<S1>:<S2>...:<S32>m` is interpreted as `\e[;m`. Needs some help here. @j4james

**Comment by j4james** [comment]
> There's still an edge case I need to figure out, `\e[1:<S1>:<S2>...:<S32>m` (first parameter overflowed the limit) shouldn't be interpreted as `\e[m`.

Only saw this after I wrote the message above, but I think my proposal for resetting the sub parameter limit would fix this too.

**Comment by tusharsnn** [comment]
I wanna work a little bit on ergonomics for accessing sub parameters. As it is, to get sub parameter at index `0` from the `std::span` of sub parameters, we would need to do `til::at(subParams, idx)`. It'll be better if we can reduce it to `subParams.at(idx)` like how we do for parameters.

I'm thinking of introducing `VTSubParameters` class for this purpose, and later update `VTParameters` to hold this instead of plain `std::span<const VTParameter>`.

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: This is addressed now. We simply parsed the above sequence as `"\e[38:2::128:0:128:1mMagenta?\e[m\n"` and this should be interpreted properly when we do support colon based Truecolor'*ing*.

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: After the discussion with @j4james, We let the parser to parse `\x1b[:m` without entering into `CsiIgnore`.. It is interpreted as `\x1b[<empty-parameter>:<empty-subparameter>m`, which makes it a no-op.

**Review by j4james — APPROVED** [review]
I've tried all the edge cases I can think of and it all looks good to me. Just a couple of nits, but you're welcome to ignore those. I appreciate you putting up with all my little technical complaints.

File: src/terminal/adapter/DispatchTypes.hpp
Comment (tusharsnn) [reply src/terminal/adapter/DispatchTypes.hpp]: No, we don't need it anymore. It is removed now.

File: src/terminal/parser/stateMachine.cpp
Comment (tusharsnn) [reply src/terminal/parser/stateMachine.cpp]: Done.

**Comment by tusharsnn** [comment]
Force pushing to check if test failed for some random reasons 😕 .

**Comment by tusharsnn** [comment]
> ##[error]The agent did not connect within the alloted time of 45 minute(s).

Possibly some internal problem. Last time it aborted the whole test, but this time only one job failed. If someone could re run the test, It'd be great :)

File: src/terminal/parser/stateMachine.hpp
Comment (lhecker) [inline src/terminal/parser/stateMachine.hpp:205]: I've tested this PR and it works great. Thank you for your hard work! I have just one question before I approve it:
Why don't we just store the subparameter range right inside the `VTParameter` struct? That way we don't need this extra vector and the parameter slicing would be a bit simpler.

Comment (tusharsnn) [reply src/terminal/parser/stateMachine.hpp:205]: `VTParameter` is base for both the `_parameter` and `_subParameter`. Storing subparameter range within VTParameter might make sense for `_parameter` but its should be illegal for `_subParameter`.

Comment (lhecker) [reply src/terminal/parser/stateMachine.hpp:205]: I personally think it's fine to store the range for both and leave it unused for sub-parameters, because I suspect that it'll not cause serious bugs. For instance, incorrectly asking a sub-param for its sub-params in turn (= illegal) could just say that there are no sub-params available (+ `abort()` in debug builds). I'm not sure if that's better than your current approach, however. It was an idea that I had when I saw the `VTParameters::subspan` function and wondered if storing the range right inside `VTParameters` would make it simpler.

For me it's a ✅ anyways and I'll leave it up to you what you personally like best. We can still tune this code however we want in the future after all. 🙂

I'll ask @DHowett to give this just a quick look, maybe he got anything else in mind.

**Review by lhecker — APPROVED** [review]

File: src/terminal/parser/OutputStateMachineEngine.cpp
Comment (DHowett) [inline src/terminal/parser/OutputStateMachineEngine.cpp:1121]: This means that `DEC?ARA` only accept subParams for index 4? nice!

File: src/terminal/adapter/DispatchTypes.hpp
Comment (DHowett) [inline src/terminal/adapter/DispatchTypes.hpp:245]: Talking myself through this:

```
   PARAMS  0                    1
   PARAMS 38 : _ : _ : _ : _ ; 48 : _ : _

SUBPARAMS       0   1   2   3        4   5
SUBPARAMS  __ : 1 : 2 : 3 : 4 ; __ : 1 : 2

(Empty spots aren't present, just used as a visual aid)

Parameter 0 has subparameter range [0,3]
Parameter 1 has subparameter range [4,5]
```

When you subspan(), we take another reference to the full subparams span but a sub-span reference to the range span?

Very clever. I love it.

**Review by DHowett — APPROVED** [review]
This is amazing, thank you. @j4james, thank you for the review and everything -- it looks like it came together so well thanks to the two of you.

Y'all, this is so cool.

**Comment by tusharsnn** [comment]
It was a great experience working on Windows Terminal for the first time ❤️, feels like an achievement to me. @j4james @zadjii-msft @DHowett @lhecker @carlos-zamora Thank you all for reviewing the PR.

**Pull request merged**
